hudi-common/src/test/java/org/apache/hudi/common/testutils/HoodieAdaptablePayloadDataGenerator.java

```diff
@@ -0,0 +1,228 @@
+    }).collect(Collectors.toList());
+  }
+
+  private static GenericRecord getInsert(int id, String pt, long ts, RecordGen recordGen) {
+    GenericRecord r = new GenericData.Record(SCHEMA);
+    r.put("id", id);
+    r.put("pt", pt);
+    return recordGen.populateForInsert(r, ts);
+  }
+
+  public static List<HoodieRecord> getUpdates(List<HoodieRecord> baseRecords, long ts, Class<?> payloadClass) throws IOException {
+    RecordGen recordGen = new RecordGen(payloadClass);
+    List<HoodieRecord> updates = new ArrayList<>();
+    Properties props = new Properties();
+    for (HoodieRecord r : baseRecords) {
+      GenericRecord gr = (GenericRecord) r.toIndexedRecord(SCHEMA, props).get().getData();
+      GenericRecord updated = getUpdate(Integer.parseInt(gr.get("id").toString()), gr.get("pt").toString(), ts, recordGen);
+      updates.add(getHoodieRecord(updated, recordGen.getPayloadClass()));
+    }
+    return updates;
+  }
+
+  private static GenericRecord getUpdate(int id, String pt, long ts, RecordGen recordGen) {
+    GenericRecord r = new GenericData.Record(SCHEMA);
+    r.put("id", id);
+    r.put("pt", pt);
+    return recordGen.populateForUpdate(r, ts);
+  }
+
+  public static List<HoodieRecord> getDeletes(List<HoodieRecord> baseRecords, long ts, Class<?> payloadClass) throws IOException {
+    RecordGen recordGen = new RecordGen(payloadClass);
+    List<HoodieRecord> deletes = new ArrayList<>();
+    Properties props = new Properties();
+    for (HoodieRecord r : baseRecords) {
+      GenericRecord gr = (GenericRecord) r.toIndexedRecord(SCHEMA, props).get().getData();
+      GenericRecord deleted = getDelete(Integer.parseInt(gr.get("id").toString()), gr.get("pt").toString(), ts, recordGen);
+      deletes.add(getHoodieRecord(deleted, recordGen.getPayloadClass()));
+    }
+    return deletes;
+  }
+
+  private static GenericRecord getDelete(int id, String pt, long ts, RecordGen recordGen) {
+    GenericRecord r = new GenericData.Record(SCHEMA);
+    r.put("id", id);
+    r.put("pt", pt);
+    return recordGen.populateForDelete(r, ts);
+  }
+
+  private static HoodieRecord getHoodieRecord(GenericRecord r, Class<?> payloadClass) throws IOException {
+    return new HoodieAvroIndexedRecord(r)
+        .prependMetaFields(
+            SCHEMA,
+            SCHEMA_WITH_METAFIELDS,
+            new MetadataValues().setRecordKey(r.get("id").toString()).setPartitionPath(r.get("pt").toString()),
+            new Properties())
+        .wrapIntoHoodieRecordPayloadWithParams(
+            SCHEMA_WITH_METAFIELDS,
+            getPayloadProps(payloadClass),
+            Option.empty(),
+            false,
+            Option.empty(),
+            false,
+            Option.of(SCHEMA));
+  }
+
+  public static class RecordGen {
+
+    public static final Set<Class<?>> SUPPORTED_PAYLOAD_CLASSES = new HashSet<>(Arrays.asList(
+        OverwriteWithLatestAvroPayload.class,
+        OverwriteNonDefaultsWithLatestAvroPayload.class,
```

Comment:
actual update for OverwriteNonDefaultsWithLatestAvroPayload, PartialUpdateAvroPayload might be different. but we should be good for the purpose of this patch. we can try to leverage community if someone really wants to write tests using how real payload might look like.